Experiment with IteratingCallback

gregw opened the pull request:
The previous semantic of `onCompleteFailure` has been renamed to `onFailure(Throwable)`, which is called immediately (but serialized) on either an abort or a failure.   A new `onCompleteFailure(Throwable)` method has been added that is called only after a `failed(throwable)` or a `abort(Throwable)` followed by `succeeded()` or `failed(Throwable)``

No usage has yet been made of the new `onCompleteFailure`, but the ICB implementation has been completely replaced by the one developed in #11876

gregw committed:
Experiment with IteratingCallback

The previous semantic of `onCompleteFailure` has been renamed to `onFailure(Throwable)`, which is called immediately (but serialized) on either an abort or a failure.   A new `onCompleteFailure(Throwable)` method has been added that is called only after a `failed(throwable)` or a `abort(Throwable)` followed by `succeeded()` or `failed(Throwable)``

No usage has yet been made of the new `onCompleteFailure`, but the ICB implementation has been completely replaced by the one developed in #11876

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:
```diff
@@ -41,20 +44,27 @@
  * Subclasses must implement method {@link #process()} where the
  * asynchronous sub-task is initiated and a suitable {@link Action}
- * is returned to this callback to indicate the overall progress of
+ * is returned to this callback to indicate the overall progress ofk
```
Typo!

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:
```diff
@@ -41,20 +44,27 @@
  * {@link #failed(Throwable)} on this callback represents the
  * completion with a failure of the large asynchronous task.
+ * <p>
+ * For most purposes, the {@link #succeeded()} and {@link #failed(Throwable)}
```
Don't we want to make them `final` for 12.1? I really see no reason to leave a dangerous possibility open.

Plus, `AsyncProxyServlet.StreamReader` overrides `failed()` which gives a compilation error as `succeeded` and `failed` are marked final.

gregw replied:
They are already final... just old javadoc.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:
```diff
@@ -198,6 +218,50 @@ protected void onCompleteFailure(Throwable cause)
+     * @param causeOrNull the cause of any {@link #abort(Throwable) abort} or {@link #failed(Throwable) failure},
+     * else {@code null} for {@link #succeeded() success}.
+     */
+    protected void onCompleted(Throwable causeOrNull)
```
The fact that you can override this method as well as `onCompleteSuccess` and `onCompleteFailure`, plus the fact that you must no forget to call super is a bit confusing, especially since ordering isn't specified and depends on the call to super.

gregw replied:
I originally had ICB explicitly calling onCF or onCS directly.  @sbordet asked for this change to have just onC wired into ICB and have it call onCF or onCS.      I can see value both ways... so you two need to pick one.

I think the javadoc is clear on the ordering as it tells us that onC calls onCS or onCF.

sbordet replied:
I'm fine with the smallest change at this point, so make this method private, and if necessary, let's open it up in the future.

lorban reviewed (dismissed):
IMHO the `onCompleteXyz` trio should be improved to leave no trace to confusion in the API. Otherwise LGTM.

gregw committed:
Experiment with IteratingCallback

The previous semantic of `onCompleteFailure` has been renamed to `onFailure(Throwable)`, which is called immediately (but serialized) on either an abort or a failure.   A new `onCompleteFailure(Throwable)` method has been added that is called only after a `failed(throwable)` or a `abort(Throwable)` followed by `succeeded()` or `failed(Throwable)``

No usage has yet been made of the new `onCompleteFailure`, but the ICB implementation has been completely replaced by the one developed in #11876

gregw committed:
Experiment with IteratingCallback

The previous semantic of `onCompleteFailure` has been renamed to `onFailure(Throwable)`, which is called immediately (but serialized) on either an abort or a failure.   A new `onCompleteFailure(Throwable)` method has been added that is called only after a `failed(throwable)` or a `abort(Throwable)` followed by `succeeded()` or `failed(Throwable)``

No usage has yet been made of the new `onCompleteFailure`, but the ICB implementation has been completely replaced by the one developed in #11876

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:
```diff
@@ -198,6 +213,93 @@ protected void onCompleteFailure(Throwable cause)
+
+    private void doOnSuccessProcessing()
+    {
+        ExceptionUtil.callAndThen(this::onSuccess, this::processing);
```
@lorban @sbordet Should I inline this call to avoid capturing `this` twice?   This is the hot path, the calls below are all error paths, but maybe we should inline them all?

sbordet replied:
Yes make `this::onSuccess` a field so it's captured once.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:
```diff
@@ -179,8 +174,24 @@ protected void onSuccess()
+     *
+     * @param cause The cause of the failure or abort
+     */
+    protected void onFailure(Throwable cause)
```
This method is called before completion, so it is wrong that it has replaced `onCompleteFailure()` almost everywhere, as most `onCompleteFailure()` implementations perform some buffer release (or similar) that are explicitly forbidden in this method.

Also, I'm not sure when could be safe to perform something from this method, because e.g. failing a callback from here may possibly result in re-entering, or calling some shared code that may assume that the buffer is released, or similar -- we call code that we don't know what it does.

gregw replied:
@sbordet The intention was to not change behaviour in this branch, as we tried that in #11876 and there were too many changes and we ended up playing whack-a-mole with bugs.    
But I'll create another branch without the renaming, to see what breaks.... my preference is still to progress with smaller PRs that fix one thing at a time, rather than changing every ICB usage (semantically) in one huge PR

gregw replied:
I have created #12055 to test this

sbordet reviewed (changes requested):
I think this is way too fine-grained.

I would be fine to "delay" the call to `onCompleteFailure()` until completeness (rather than early failure), and that's it.

The new methods introduced are currently not used, and will be APIs that we need to maintain and carefully document about what can and cannot be done in them, which I think is a big effort not worth it.
Their semantic is delicate and we got it mostly wrong in this PR (e.g. calling release from onFailure()).

So I'd revert to minimal changes, i.e. just "delay" properly the call to `onCompleteFailure()`, and all the rest should just be correct as it is.

gregw committed:
Reduce capture of this:: on hot path

sbordet committed:
Split the release operations in case of failures.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Made onCompleted() private to address PR review.

This is necessary to avoid confusion, for example class A_ICB extending ICB and overriding onCompleted() without calling super, and class B_ICB extending A_ICB and overriding onCompleteSuccess() which would be never called.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@gregw I think it is mandatory that "release" operations in case of failure become in fact "remove" operations, when dealing with buffers, because it is always possible that the failure is asynchronous, and the buffer is being utilized by the async operation while the release is trying to put it back into the pool for reuse.
It cannot be reused.

Therefore, rather than splitting the "fail" and "release" operations, the former in `onFailure()` and the latter in `onCompleteFailure()` (which I have done as an experiment in in branch `experiment/jetty-12.1.x/IteratingCallback-SplitRelease`), we should instead do the "fail" and "remove" actions from `onFailure()`.

Which then begs the question if we need to have the additional methods introduced in this PR especially regarding abort/failure.
The reason is that current code overriding `onCompleteFailure()` would have a different behavior with this PR, possibly breaking users like #12055 showed.

So the idea is to have `onCompleteFailure()` behave like it always did, but fix the "release" operations in case of failure with "remove" operations, and not add the new methods such as `onAborted()`, `onFailed()`, etc.

Or, we bit the bullet, change the behavior in 12.1.x, introduce the new methods, possibly break existing code, but avoid the split of my experiment PR by introducing "remove" semantic to RBB.

gregw commented:
@lorban Since #12055 has been closed by @sbordet, can we review this one.    The aim here is to provide the option for the new semantic, with better naming, but to not change behaviour.    Behavior changes can be done bit by bit in subsequent PRs.

gregw committed:
Merge branch 'jetty-12.1.x' into experiment/jetty-12.1.x/IteratingCallback

# Conflicts:
#	jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponseAndCallback.java

gregw committed:
Fix

gregw committed:
Merge branch 'jetty-12.1.x' into experiment/jetty-12.1.x/IteratingCallback

gregw committed:
reverted import changes from bad merge

gregw committed:
Merge branch 'jetty-12.1.x' into experiment/jetty-12.1.x/IteratingCallback

gregw commented:
@lorban can you take this one on?

sbordet commented:
@gregw a note to self/team about failures in case of non termination.

The practical example would be a write that is TCP congested, and a timeout happening.

`onFailure()` would be called, but the write won't complete unless TCP congestion resolves, or the endPoint is closed in `onFailure()` (I assume that closing the endPoint would fail the write, but needs to be tested).

gregw commented:
> @gregw a note to self/team about failures in case of non termination.
> 
> The practical example would be a write that is TCP congested, and a timeout happening.
> 
> `onFailure()` would be called, but the write won't complete unless TCP congestion resolves, or the endPoint is closed in `onFailure()` (I assume that closing the endPoint would fail the write, but needs to be tested).

@sbordet  ... and further to that  example, we have two strategies to deal with the scenario:

 a) in `onFailure` we remove the write buffer from the pooling, so that it will not be reused ever.
 b) we move releasing the write the write buffer to `onCompleteFailure`, so it is reused, but only once the congestion is resolved and the write completes (or fails)

The buffer is held in memory in both a) and b), so there is no memory saving in either.    In a) we potentially free a slot in the buffer pool making for better pooling for other requests, but only a very marginal gain.    So the choice between a) and b) really should be based on code clarity and what other events might be triggered by `onFailure` and/or `onCompleteFailure`.   I think both are valid, so we should make `remove()` a first class operation on RBB

gregw committed:
Merge branch 'jetty-12.1.x' into experiment/jetty-12.1.x/IteratingCallback

gregw committed:
releaseForRemoval

gregw commented:
@sbordet as per my comment above, I have added `RetainableByteBuffer.releaseForRemoval()`

@janbartel I'm seeing a lot of differences between ee10 and ee11, I think we may have messed up some merge forwards.   I think we need to do a good compare on jetty-12.0.x to jetty-12.1.x, and then the same from EE10 to EE11

gregw committed:
improved javadoc

gregw committed:
Merge branch 'jetty-12.1.x' into experiment/jetty-12.1.x/IteratingCallback

# Conflicts:
#	documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/ContentDocs.java
#	documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/SelectorManagerDocs.java
#	documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/WebSocketDocs.java
#	documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/ServerDocs.java

gregw committed:
Merge branch 'jetty-12.1.x' into experiment/jetty-12.1.x/IteratingCallback

gregw committed:
Merge branch 'jetty-12.1.x' into experiment/jetty-12.1.x/IteratingCallback

gregw commented:
@sbordet @lorban It looks like in a lot of places we have returned `Action.IDLE` from process when we have actually scheduled demand.    I don't think that previously that mattered so much, but now we have the full completion contract, I don't IDLE is the right thing to return here as abort/close will be handled immediately without regards to the pending demand callback.    I'm fixing as many as I can....

gregw committed:
WIP

gregw commented:
@sbordet @lorban We have a lot of ICB usage that does something like:
```java
   if (!out.isReady())
       return Action.IDLE;
```

or 

```java
   if (filled == 0)
   {
       fillInterested();
       return Action.IDLE;
    }
```

or 

```java
   if (chunk == null)
   {
       demand(this);
       return Action.IDLE;
    }
```

And then the `onDataAvailable()` or `onFillable()`  or demand callback method just calls iterate.        This is wrong because the ICB is not IDLE, it is SCHEDULED and a callback is going to happen, even if the ICB is aborted or closed.

It is difficult to change these usages, because often the ICB action is initiated by an initial call(back) to `onDataAvailable()` or `onFillable`, so it is almost as if the ICB has been created in SCHEDULED state.

I'm not sure we should try to fix in this PR, but we need to look closely at all our ICBs

gregw committed:
WIP

gregw committed:
documentation

gregw committed:
Merge remote-tracking branch 'origin/experiment/jetty-12.1.x/IteratingCallback-SplitRelease' into experiment/jetty-12.1.x/IteratingCallback

# Conflicts:
#	documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/ContentDocs.java
#	jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/internal/HTTP2Flusher.java
#	jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponseAndCallback.java
#	jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpOutput.java
#	jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/HttpOutput.java
#	jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/HttpOutput.java

gregw committed:
WIP

lorban commented on documentation/jetty/modules/code/examples/src/main/java/org/eclipse/jetty/docs/programming/ContentDocs.java:
```diff
@@ -361,14 +364,19 @@ protected void onCompleteSuccess()
+        {
+            // In case of a failure, we wait until the write has completed before releasing any chunk.
+            if (chunk != null)
+                chunk.release();
```
Doing `chunk = Content.Chunk.next(chunk);` like in `onSuccess()` sounds wise.

sbordet commented on documentation/jetty/modules/code/examples/src/main/java/org/eclipse/jetty/docs/programming/SelectorManagerDocs.java:
```diff
@@ -277,7 +277,7 @@ else if (filled == 0)
+                        fillInterested(this);
 
                         // Signal that the iteration is now IDLE.
                         return Action.IDLE;
```
Should be SCHEDULED.

sbordet commented on documentation/jetty/modules/code/examples/src/main/java/org/eclipse/jetty/docs/programming/server/ServerDocs.java:
```diff
@@ -176,7 +176,7 @@ public void onOpen()
         public void onFillable()
         {
-            callback.iterate();
+            callback.succeeded();
```
This should be as before, `callback.iterate()`, because `onFillable()` is called by the `fillInterested()` call from `onOpen()`.

It's inside `process()` that we do things differently.

gregw replied:
Ah nice way to solve the difference between the  first "demand" and the subsequent demand.

sbordet commented on documentation/jetty/modules/code/examples/src/main/java/org/eclipse/jetty/docs/programming/server/ServerDocs.java:
```diff
@@ -222,8 +222,8 @@ else if (filled == 0)
                         // again interest for fill events.
                         fillInterested();
```
This should be `fillInterested(this)`.

sbordet commented on documentation/jetty/modules/programming-guide/pages/arch/io.adoc:
```diff
@@ -209,13 +209,14 @@ If this was the only active network connection, the system would now be idle, wi
 
 The last case is that during `process()` an exception is thrown, for example by `EndPoint.fill(ByteBuffer)` or, in more advanced implementations, by code that parses the bytes that have been read and finds them unacceptable; any exception thrown within `process()` will be caught by `IteratingCallback` that will exit the loop with a failure and call `onCompleteFailure(Throwable)` with the exceptio ...
 
+Note that some failures may occur whilst a scheduled operation is in progress.  Such failures are notified immediately via the `onFailure(Throwable)` method, but care must be taken to not release any resources that may still be in use by the scheduled operation. The `onCompleteFailure(Throwable)` method is called when both a failure has occurred and any scheduled operation has completed.  An exam ...
```
Every sentence in its own line.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPool.java:
```diff
@@ -63,11 +63,9 @@ public interface ByteBufferPool
      */
-    @Deprecated
-    default boolean removeAndRelease(RetainableByteBuffer buffer)
+    default boolean releaseForRemoval(RetainableByteBuffer buffer)
```
I find the name strange. How about `releaseAndRemove()` or just `remove()` with javadocs explaining that also has release semantic?

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentCopier.java:
```diff
@@ -76,14 +73,26 @@ protected Action process() throws Throwable
+    @Override
+    protected void onSuccess()
+    {
+        super.onSuccess();
```
Remove, unnecessary.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentCopier.java:
```diff
@@ -76,14 +73,26 @@ protected Action process() throws Throwable
-        ExceptionUtil.callAndThen(x, source::fail, super::onCompleteFailure);
+            if (current != null)
+                current.release();
+        }, super::onCompleteFailure);
```
No need to call `super`.

gregw committed:
updates from review

gregw committed:
Fixed FCGI Flusher

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -1080,6 +1080,27 @@ static Chunk next(Chunk chunk)
+         * @param chunk The chunk to release or {@code null}
+         * @return The {@link #next(Chunk)} chunk;
+         */
+        static Chunk releaseNext(Chunk chunk)
```
Rename to `releaseAndNext()`.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -163,11 +163,24 @@ default Mutable asMutable() throws ReadOnlyBufferException
         return release();
     }
 
+    @Override
+    default boolean release()
+    {
+        return Retainable.super.release();
+    }
+
```
Is this override necessary?

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConnectHandler.java:
```diff
@@ -736,18 +737,17 @@ protected Action process()
+                    if (filled == 0)
                     {
-                        buffer.release();
                         fillInterested();
                         return Action.IDLE;
```
Should not this be:

```java
fillInterested(this);
return SCHEDULED;
```

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/internal/HttpSenderOverHTTP.java:
```diff
@@ -334,11 +332,16 @@ protected Action process() throws Exception
-            callback.failed(cause);
         }
 
         private void release()
```
You could use `Retainable.release` here too.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -2338,6 +2380,28 @@ protected Action process()
+                        protected void onSuccess()
+                        {
+                            // release the last buffer written
+                            if (_buffer != null)
```
use `Retainable.release()`?

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java:
```diff
@@ -2338,6 +2380,28 @@ protected Action process()
+                        protected void onCompleteFailure(Throwable x)
+                        {
+                            // release the last buffer written
+                            if (_buffer != null)
```
use `Retainable.release()`?

lorban commented:
FYI `HttpClientStreamTest.testInputStreamContentProviderThrowingWhileReading()` seems to fail because of a leak introduced in FCGI.

gregw committed:
Merge branch 'jetty-12.1.x' into experiment/jetty-12.1.x/IteratingCallback

# Conflicts:
#	jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpOutput.java
#	jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/HttpOutput.java

gregw committed:
updates from review

lorban committed:
fix reported leak error messages

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix leak

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix missing calls to super

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into experiment/jetty-12.1.x/IteratingCallback

sbordet committed:
Small tweaks to documentation and javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into experiment/jetty-12.1.x/IteratingCallback

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into experiment/jetty-12.1.x/IteratingCallback

gregw committed:
deflake test

gregw commented:
@sbordet are you OK if I dismiss @lorban's review and merge this once I get a build clean of flakes?

sbordet commented:
@gregw I'm ok to merge this.

gregw commented:
The QoSHandler CI failures appear to be a problem in head, so merging this without a green build.
